File: parameters_py/parameters_py/param_monitor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import ParameterDescriptor, SetParametersResult

class MiniParam(Node):
    def __init__(self, name):
        super().__init__(name)
        # set param description
        parameter_descriptor = ParameterDescriptor(description='This parameter is mine!')

        self.declare_parameter("demo_param", "set in python !!", parameter_descriptor)

        # Monitor updates to 'demo_param'
        self.add_on_set_parameters_callback(self.demo_param_change_callback)

    def demo_param_change_callback(self, params):
        result = SetParametersResult()  # Create a result instance
        for param in params:
            if param.name == 'demo_param':
                self.get_logger().info(f'received an update to : {param.name}, type is {type(param.value).__name__}, value is {param.value}')
                result.successful = True
        return result
            
def main(args=None):
    rclpy.init(args=args)
    node = MiniParam("param_monitor_py")
    rclpy.spin(node)
    
if __name__ == '__main__':
    main()

File: parameters_cpp/src/param_monitor.cpp
#include <chrono>
#include <functional>
#include <string>

#include <rclcpp/rclcpp.hpp>

#include "rclcpp/rclcpp.hpp"

class MiniParam : public rclcpp::Node {
public:
    MiniParam() : Node("param_monitor_cpp") {
        // set param description
        auto param_desc = rcl_interfaces::msg::ParameterDescriptor{};
        param_desc.description = "This parameter is mine!";
        this->declare_parameter("demo_param", "set in cpp !!", param_desc);

        // used to monitor parameter changes
        param_sub_ = std::make_shared<rclcpp::ParameterEventHandler>(this);
        // 监测本节点的参数变化
        demo_param_change_cb_ = param_sub_->add_parameter_callback("demo_param", 
        [this](const rclcpp::Parameter &p) {
            RCLCPP_INFO(this->get_logger(), "received an update to : %s , type is %s, value is %s", 
                        p.get_name().c_str(),
                        p.get_type_name().c_str(),
                        p.as_string().c_str());
        });
        // 监测其他节点的参数变化
        turtlesim_node_background_b_change_cb_ = param_sub_->add_parameter_callback("background_b",
        [this](const rclcpp::Parameter &p) {
            RCLCPP_INFO(this->get_logger(), "received an update to : %s , type is %s, value is %ld", 
                        p.get_name().c_str(),
                        p.get_type_name().c_str(),
                        p.as_int());
        }, "/turtlesim_node");
    }

private:
    std::shared_ptr<rclcpp::ParameterEventHandler> param_sub_;
    std::shared_ptr<rclcpp::ParameterCallbackHandle> demo_param_change_cb_;
    std::shared_ptr<rclcpp::ParameterCallbackHandle> turtlesim_node_background_b_change_cb_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);

    rclcpp::spin(std::make_shared<MiniParam>());

    rclcpp::shutdown();

    return 0;
}